runners/direct-java/src/main/java/org/apache/beam/runners/direct/TestStream.java

```diff
@@ -0,0 +1,347 @@
+      this.typeDescriptor = typeDescriptor;
+      events = ImmutableList.builder();
+
+      currentWatermark = BoundedWindow.TIMESTAMP_MIN_VALUE;
+    }
+
+    /**
+     * Adds the specified elements to the source with timestamp equal to the current watermark.
+     *
+     * @return this {@link TestStream.Builder}
+     */
+    @SafeVarargs
+    public final Builder<T> addElements(
+        T element, T... elements) {
+      TimestampedValue<T> firstElement = TimestampedValue.of(element, currentWatermark);
+      @SuppressWarnings("unchecked")
+      TimestampedValue<T>[] remainingElements = new TimestampedValue[elements.length];
+      for (int i = 0; i < elements.length; i++) {
+        remainingElements[i] = TimestampedValue.of(elements[i], currentWatermark);
+      }
+      return addElements(firstElement, remainingElements);
+    }
+
+    /**
+     * Adds the specified elements to the source with the provided timestamps.
+     *
+     * @return this {@link TestStream.Builder}
+     */
+    @SafeVarargs
+    public final Builder<T> addElements(
+        TimestampedValue<T> element, TimestampedValue<T>... elements) {
+      events.add(TestSourceEvent.addElements(element, elements));
+      return this;
+    }
+
+    /**
+     * Advance the watermark of this source to the specified instant.
+     *
+     * <p>The watermark must advance monotonically and to at most
+     * {@link BoundedWindow#TIMESTAMP_MAX_VALUE}.
+     *
+     * @return this {@link TestStream.Builder}
+     */
+    public Builder<T> advanceWatermarkTo(Instant newWatermark) {
+      checkArgument(newWatermark.isAfter(currentWatermark),
+          "The watermark must monotonically advance");
+      checkArgument(newWatermark.isBefore(BoundedWindow.TIMESTAMP_MAX_VALUE),
+          "The Watermark cannot progress beyond the maximum. Got: %s. Maximum: %s",
+          newWatermark,
+          BoundedWindow.TIMESTAMP_MAX_VALUE);
+      events.add(TestSourceEvent.<T>advanceWatermarkTo(newWatermark));
+      currentWatermark = newWatermark;
+      return this;
+    }
+
+    /**
+     * Advance the processing time by the specified amount.
+     *
+     * @return this {@link TestStream.Builder}
+     */
+    public Builder<T> advanceProcessingTime(Duration amount) {
+      checkArgument(amount.getMillis() > 0,
+          "Must advance the processing time by a positive amount. Got: ",
+          amount);
+      events.add(TestSourceEvent.<T>advanceProcessingTime(amount));
+      return this;
+    }
+
+    /**
+     * Advance the watermark to infinity, completing this {@link TestStream}. Future calls to
+     * the same builder will not affect the returned {@link TestStream}.
+     */
+    public TestStream<T> advanceWatermarkToInfinity() {
+      events.add(TestSourceEvent.<T>advanceWatermarkTo(BoundedWindow.TIMESTAMP_MAX_VALUE));
+      return new TestStream<>(typeDescriptor, events.build());
+    }
+  }
+
+  @AutoValue
+  abstract static class TestSourceEvent<T> {
```

Comment:
We should probably have three different types of events (sub-classes) since having one with all the different methods adds some confusion as to whether two or more of the properties can be set.

If you only expose the one field at a time by subclassing, the implementor will know exactly what they must do for each event.

Also, allows for a runner to scan the list and see if I support all the event types and potentially add support for more event types in the future.